### PR TITLE
Terraform state backup and restore tool

### LINKED ISSUE #310: Backup and restore the Terraform state
We can build a command line tool to `backeup` and `restore` the Configuration Object and the Terraform state. The command line tool may have two subcommands: `backup` and `restore`.

The usage of the `backup` subcommand may like this:

```
$ tfc backup --help
Backup the Configuration Object, and if the configuration use the default Terraform kubernetes backend, backup the Terraform state too

Usage:
  tfc backup [OPTIONS] CONFIGURATION_NAME

Examples:

  tfc backup -d ./backup_dir configuration-oss-demo


Flags:
  -d, --dir string   Destination directory for saving the backed up files
  -h, --help         help for backup

Global Flags:
      --as string                      Username to impersonate for the operation
      --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
      --cache-dir string               Default cache directory (default "/Users/loheagn/.kube/cache")
      --certificate-authority string   Path to a cert file for the certificate authority
      --client-certificate string      Path to a client certificate file for TLS
      --client-key string              Path to a client key file for TLS
      --cluster string                 The name of the kubeconfig cluster to use
      --context string                 The name of the kubeconfig context to use
      --insecure-skip-tls-verify       If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure
      --kubeconfig string              Path to the kubeconfig file to use for CLI requests.
  -n, --namespace string               If present, the namespace scope for this CLI request
      --password string                Password for basic authentication to the API server
      --request-timeout string         The length of time to wait before giving up on a single server request. Non-zero values should contain a corresponding time unit (e.g. 1s, 2m, 3h). A value of zero means don't timeout requests. (default "0")
  -s, --server string                  The address and port of the Kubernetes API server
      --tls-server-name string         Server name to use for server certificate validation. If it is not provided, the hostname used to contact the server is used
      --token string                   Bearer token for authentication to the API server
      --user string                    The name of the kubeconfig user to use
      --username string                Username for basic authentication to the API server
```

The `restore` subcommand can be used to restore the Configuration Object and it will also restore the Terraform state if the configuration uses the default Terraform kubernetes backend.

When the `retore` subcommand works, it will accept a yaml file or dir. It fetches the configuration definations from the yaml file or the dir, and check if the configurations use the default kubernetes backend, if the configuration does, the `restore` subcommand will read the `state.json` file and restore the bakcend secret before resotring the configuration objects.

The usage of the `restore` subcommand may like this:

```
$ tfc restore --help
Restore the Configuration Object, and if the configuration use the default Terraform kubernetes backend, restore the Terraform state too

Usage:
  tfc restore OPTIONS

Examples:

  tfc restore -f ./configuration.yaml


Flags:
  -f, --from string   A file or a directory from where the command read the source definitions
  -h, --help          help for restore

Global Flags:
      --as string                      Username to impersonate for the operation
      --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
      --cache-dir string               Default cache directory (default "/Users/loheagn/.kube/cache")
      --certificate-authority string   Path to a cert file for the certificate authority
      --client-certificate string      Path to a client certificate file for TLS
      --client-key string              Path to a client key file for TLS
      --cluster string                 The name of the kubeconfig cluster to use
      --context string                 The name of the kubeconfig context to use
      --insecure-skip-tls-verify       If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure
      --kubeconfig string              Path to the kubeconfig file to use for CLI requests.
  -n, --namespace string               If present, the namespace scope for this CLI request
      --password string                Password for basic authentication to the API server
      --request-timeout string         The length of time to wait before giving up on a single server request. Non-zero values should contain a corresponding time unit (e.g. 1s, 2m, 3h). A value of zero means don't timeout requests. (default "0")
  -s, --server string                  The address and port of the Kubernetes API server
      --tls-server-name string         Server name to use for server certificate validation. If it is not provided, the hostname used to contact the server is used
      --token string                   Bearer token for authentication to the API server
      --user string                    The name of the kubeconfig user to use
      --username string                Username for basic authentication to the API server
```

### PR DESCRIPTION
**Pull request by loheagn**
This PR fixes #310 

Signed-off-by: loheagn <loheagn@icloud.com>

### CODE REVIEW

File: hack/tool/backup_restore/cmd/restore.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+			if err != nil {
+				return err
+			}
+			stateJSONPath = filepath.Join(pwd, stateJSONPath)
+			configurationPath = filepath.Join(pwd, configurationPath)
+			return restore(context.Background())
+		},
+	}
+	restoreCmd.Flags().StringVar(&stateJSONPath, "state", "state.back.json", "the path of the backed up Terraform state file")
+	restoreCmd.Flags().StringVar(&configurationPath, "configuration", "configuration.back.yaml", "the path of the backed up configuration objcet yaml file")
```
Comment (zzxwill) [inline hack/tool/backup_restore/cmd/restore.go:66]:
```suggestion
	restoreCmd.Flags().StringVar(&configurationPath, "configuration", "configuration.yaml", "the path of the backed up configuration objcet yaml file")
```
Normally the configuration file is there, no need to have a backup.

File: hack/tool/backup_restore/cmd/restore.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+				"tfstate": tfState,
+			},
+		}
+		if err := k8sClient.Create(ctx, &gotSecret); err != nil {
+			return err
+		}
+	} else {
+		// update the secret
+		gotSecret.Annotations["encoding"] = "gzip"
+		gotSecret.Labels["app.kubernetes.io/managed-by"] = "terraform"
```
Comment (zzxwill) [inline hack/tool/backup_restore/cmd/restore.go:147]: Are they necessary?

Comment (loheagn) [reply hack/tool/backup_restore/cmd/restore.go:147]: I'm not sure. But the secret created by the Terraform contains these annotations and labels. I think we'd better add them for consistency.

File: hack/tool/backup_restore/README.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
+# Backup & Restore
+
+This go module is a command line tool to back up and restore the configuration and the Terraform state.
+
+It has two subcommands `backup` and `restore`.
+
+## `restore`
+
+The main usage of the `restore` subcommand is to import an "outside" Terraform instance (maybe created by the terraform command line or managed by another terraform-controller before) to the terraform-controller in the target kubernetes without recreating the cloud resources.
```
Comment (zzxwill) [inline hack/tool/backup_restore/README.md:9]:
```suggestion
The main usage of the `restore` subcommand is to import an "outside" Terraform instance (maybe created by the terraform command line or managed by another terraform-controller before) to the terraform-controller in the target Kubernetes without recreating the cloud resources.
```

File: hack/tool/backup_restore/README.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+To use the `restore` subcommand, you should:
+
+1. Prepare the `configuration.yaml` file and the `state.json` file whose content is the state json of the Terraform instance which would be imported.
+
```
Comment (zzxwill) [inline hack/tool/backup_restore/README.md:14]: This is a typical scenario, great! Here is another one, please cover it in this PR or in a later one https://github.com/kubevela/terraform-controller/issues/310#issuecomment-1133548163.

Comment (zzxwill) [reply hack/tool/backup_restore/README.md:14]: Is this should be covered by `backup` command?

Comment (loheagn) [reply hack/tool/backup_restore/README.md:14]: I can add the `backup` command in the next PR.

**Review by zzxwill — APPROVED** [review]
Good job!

**Pull request merged**